src/Stratis.Bitcoin/Consensus/PeerGetHeaderRequestBehavior.cs

```diff
@@ -0,0 +1,149 @@
+
+        protected override void AttachCore()
+        {
+            this.AttachedPeer.MessageReceived.Register(this.OnMessageReceived, true);
+        }
+
+        protected override void DetachCore()
+        {
+            this.AttachedPeer.MessageReceived.Unregister(this.OnMessageReceived);
+        }
+
+        /// <summary>
+        /// Processes an incoming message from the peer.
+        /// </summary>
+        /// <param name="peer">Peer from which the message was received.</param>
+        /// <param name="message">Received message to process.</param>
+        private Task OnMessageReceived(INetworkPeer peer, IncomingMessage message)
+        {
+            switch (message.Message.Payload)
+            {
+                case GetHeadersPayload getHeaders:
+                    this.HandleGetHeaders(getHeaders);
+                    break;
+
+                default:
+                    break;
+            }
+
+            return Task.CompletedTask;
+        }
+
+        /// <summary>
+        /// Determines whether or not a peer asked for the same set of headers within 60 seconds.
+        /// <para>
+        /// If the same set of headers was requested more than <see cref="GetHeaderRequestCountThreshold"/>, it will be banned
+        /// and disconnected.
+        /// </para>
+        /// </summary>
+        private void HandleGetHeaders(GetHeadersPayload getHeaders)
+        {
+            var blockLocatorHash = getHeaders.BlockLocator.Blocks.FirstOrDefault();
+            if (blockLocatorHash == null)
+            {
+                this.logger.LogTrace("(-)[EMPTY_BLOCKLOCATOR]");
+                return;
+            }
+
+            // Is the last requested hash the same as this request.
+            if (this.getHeaderLastRequestHash == blockLocatorHash)
+            {
+                this.logger.LogDebug($"{this.AttachedPeer.PeerEndPoint} block locator matches previous, count {this.getHeaderRequestCount}");
+
+                // Was this hash requested less than 60 seconds ago.
+                if (this.getHeaderLastRequestedTimestamp > this.dateTimeProvider.GetUtcNow().AddSeconds(-60))
+                {
+                    this.getHeaderRequestCount++;
+                }
+                else
+                {
+                    this.getHeaderLastRequestedTimestamp = this.dateTimeProvider.GetUtcNow();
+                    this.getHeaderRequestCount = 0;
+
+                    this.logger.LogTrace("(-)[LAST_REQUESTED_WINDOW_ELAPSED]");
+                    return;
+                }
+
+                // If the same header was requested more than 3 times in the last 60 seconds,
+                // ban and disconnect the peer for 1 hour.
+                if (this.getHeaderRequestCount >= GetHeaderRequestCountThreshold)
+                {
+                    this.peerBanning.BanAndDisconnectPeer(this.AttachedPeer.PeerEndPoint, BanDurationSeconds, $"Banned via {this.GetType().Name} for {BanDurationSeconds} seconds.");
```

Comment:
any valid reason for using `this.GetType().Name` instead of just PeerGetHeaderRequestBehavior?

Reply:
Ah actually maybe remove the GetType all together just put some text there
'Banned via rate limiting for X seconds'